Bottom: 3f598a35fdae07d67768a46a76a6c291b949b51f
Top:    fd9a74e7c8ac4e668106af7280aaa948ff7c97ae
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2011-04-28 22:38:43 +0100

Refresh of fix-livedraw-to-work-with-norm

---

diff --git a/src/autoroute.c b/src/autoroute.c
index 93a0e3d..01bdfa0 100644
--- a/src/autoroute.c
+++ b/src/autoroute.c
@@ -4493,6 +4493,30 @@ no_expansion_boxes (routedata_t * rd)
 }
 #endif
 
+static void
+ripout_livedraw_obj (routebox_t *rb)
+{
+  if (rb->type == LINE && rb->livedraw_obj.line)
+    {
+      LayerType *layer = LAYER_PTR (PCB->LayerGroups.Entries[rb->group][0]);
+      EraseLine (rb->livedraw_obj.line);
+      DestroyObject (PCB->Data, LINE_TYPE, layer, rb->livedraw_obj.line, NULL);
+    }
+  if (rb->type == VIA && rb->livedraw_obj.via)
+    {
+      EraseVia (rb->livedraw_obj.via);
+      DestroyObject (PCB->Data, VIA_TYPE, rb->livedraw_obj.via, NULL, NULL);
+    }
+}
+
+static int
+ripout_livedraw_obj_cb (const BoxType * b, void *cl)
+{
+  routebox_t *box = (routebox_t *) b;
+  ripout_livedraw_obj (box);
+  return 0;
+}
+
 struct routeall_status
 {
   /* --- for completion rate statistics ---- */
@@ -4612,20 +4636,7 @@ RouteAll (routedata_t * rd)
 		  if (rip)
 		    {
                       if (TEST_FLAG (LIVEROUTEFLAG, PCB))
-                        {
-                          if (p->type == LINE && p->livedraw_obj.line)
-                            {
-                              LayerType *layer =
-                                LAYER_PTR (PCB->LayerGroups.Entries[p->group][0]);
-                              EraseLine (p->livedraw_obj.line);
-                              DestroyObject (PCB->Data, LINE_TYPE, layer, p->livedraw_obj.line, NULL);
-                            }
-                          if (p->type == VIA && p->livedraw_obj.via)
-                            {
-                              EraseVia (p->livedraw_obj.via);
-                              DestroyObject (PCB->Data, VIA_TYPE, p->livedraw_obj.via, NULL, NULL);
-                            }
-                        }
+                        ripout_livedraw_obj (p);
 		      del =
 			r_delete_entry (rd->layergrouptree[p->group],
 					&p->box);
@@ -5150,6 +5161,19 @@ AutoRoute (bool selected)
   /* auto-route all nets */
   changed = (RouteAll (rd).total_nets_routed > 0) || changed;
 donerouting:
+  if (changed && TEST_FLAG (LIVEROUTEFLAG, PCB))
+    {
+      int i;
+      BoxType big;
+      big.X1 = 0;
+      big.X2 = MAX_COORD;
+      big.Y1 = 0;
+      big.Y2 = MAX_COORD;
+      for (i = 0; i < max_group; i++)
+        {
+          r_search (rd->layergrouptree[i], &big, NULL, ripout_livedraw_obj_cb, NULL);
+        }
+    }
   if (changed)
     changed = IronDownAllUnfixedPaths (rd);
   Message ("Total added wire length = %f inches, %d vias added\n",
